src/stream/src/task/lru_manager.rs

```diff
@@ -0,0 +1,341 @@
+        &self.inner
+    }
+}
+
+impl<K, V, S, A: Clone + Allocator> DerefMut for ManagedLruCache<K, V, S, A> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+pub struct LruManager {
+    watermark_epoch: Arc<AtomicU64>,
+    total_memory_available_bytes: usize,
+    barrier_interval_ms: u32,
+}
+
+pub type LruManagerRef = Arc<LruManager>;
+
+pub enum ExecutorCache<K, V, S = DefaultHasher, A: Clone + Allocator = Global> {
+    /// An managed cache. Eviction depends on the node memory usage.
+    Managed(ManagedLruCache<K, V, S, A>),
+    /// An local cache. Eviction depends on local executor cache limit setting.
+    Local(EvictableHashMap<K, V, S, A>),
+}
+
+impl<K: Hash + Eq, V, S: BuildHasher, A: Clone + Allocator> ExecutorCache<K, V, S, A> {
+    /// Evict epochs lower than the watermark
+    pub fn evict(&mut self) {
+        match self {
+            ExecutorCache::Managed(cache) => cache.evict(),
+            ExecutorCache::Local(cache) => cache.evict_to_target_cap(),
+        }
+    }
+
+    /// An iterator visiting all values in most-recently used order. The iterator element type is
+    /// &V.
+    #[auto_enum(Iterator)]
+    pub fn values(&self) -> impl Iterator<Item = &V> {
+        match self {
+            ExecutorCache::Managed(cache) => cache.iter().map(|(_k, v)| v),
+            ExecutorCache::Local(cache) => cache.iter().map(|(_k, v)| v),
+        }
+    }
+
+    /// An iterator visiting all values mutably in most-recently used order. The iterator element
+    /// type is &mut V.
+    #[auto_enum(Iterator)]
+    pub fn values_mut(&mut self) -> impl Iterator<Item = &mut V> {
+        match self {
+            ExecutorCache::Managed(cache) => cache.iter_mut().map(|(_k, v)| v),
+            ExecutorCache::Local(cache) => cache.iter_mut().map(|(_k, v)| v),
+        }
+    }
+}
+
+impl<K, V, S, A: Clone + Allocator> Deref for ExecutorCache<K, V, S, A> {
+    type Target = LruCache<K, V, S, A>;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            ExecutorCache::Managed(cache) => &cache.inner,
+            ExecutorCache::Local(cache) => &cache.inner,
+        }
+    }
+}
+
+impl<K, V, S, A: Clone + Allocator> DerefMut for ExecutorCache<K, V, S, A> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        match self {
+            ExecutorCache::Managed(cache) => &mut cache.inner,
+            ExecutorCache::Local(cache) => &mut cache.inner,
+        }
+    }
+}
+
+impl LruManager {
+    const EVICTION_THRESHOLD_AGGRESSIVE: f64 = 0.9;
+    const EVICTION_THRESHOLD_GRACEFUL: f64 = 0.7f64;
+
+    pub fn new(total_memory_available_bytes: usize, barrier_interval_ms: u32) -> Arc<Self> {
+        let manager = Arc::new(Self {
+            watermark_epoch: Arc::new(0.into()),
+            total_memory_available_bytes,
+            barrier_interval_ms,
+        });
+
+        // Run a background memory monitor
+        tokio::spawn(manager.clone().run());
```

Comment:
It would be better not to bring some side effects in `new`. Naming as `Manager`, let's manually `run` it. 😄